### ISSUE TITLE
Problems with using libsndfile & scipy

### ISSUE THREAD

**Issue by jcelerier** [scipy/scipy#8130]
I was investigating a weird crash that I had in a software that uses SciPy and links to libsndfile.

The backtrace looks like this: 

    #0  0x00007fffe281936f in sf_error () from /usr/lib/libsndfile.so.1
    #1  0x00007fffda890db6 in sf_error_check_fpe (func_name=0x7fffda99a2e5 "expit") at scipy/special/sf_error.c:152

I couldn't see what was wrong for a good hour but the problem is evident: there are two `sf_error` functions being loaded at the same time:

    https://github.com/scipy/scipy/blob/master/scipy/special/sf_error.c#L52

and 

    https://github.com/erikd/libsndfile/blob/fa1d42fdddcee203c7b0348324f9e657a9eb9c5a/src/sndfile.c#L494

obviously, `sf_error_check_fpe` calls an unexpected `sf_error` with a wildly different ABI which leads to such crashes. What would be easy ways to prevent this ?

**Comment by pv** [comment]
I think this shouldn't normally occur.

Does the "software that uses SciPy" use an embedded Python interpreter? 
Does it load libraries with RTLD_GLOBAL enabled?

**Comment by jcelerier** [comment]
>  Does the "software that uses SciPy" use an embedded Python interpreter? 

yep

> Does it load libraries with RTLD_GLOBAL enabled?

mhhh... it uses Qt's plug-in system, that's a possibility.

**Comment by jcelerier** [comment]
Here's hwo to reproduce (I use [pybind11](https://github.com/pybind/pybind11) ) : 

CMakeLists.txt : 
```
set(CMAKE_BUILD_TYPE Debug)
cmake_minimum_required(VERSION 3.1)
project(blah)
add_subdirectory(pybind11)

add_executable(main main.cpp)
target_link_libraries(main dl pybind11::embed pthread avdevice)
```

main.cpp: 

```
#include <pybind11/embed.h>
int main()
{
  pybind11::scoped_interpreter guard{};
  pybind11::exec("import sklearn");
}
```

**Comment by pv** [comment]
Doesn't seem to repro on ubuntu 17.10

**Comment by pv** [comment]
Does something mess with sys.setdlopenflags. The following passes here:
```
import sys, os, sklearn
assert sys.getdlopenflags() == os.RTLD_NOW
```

**Comment by jcelerier** [comment]
hmmm, it seems that on ubuntu, libavdevice does not link directly with libsndfile.
Here's a repro on ubuntu 17.10:

```CMake
set(CMAKE_BUILD_TYPE Debug)
cmake_minimum_required(VERSION 3.1)
project(blah)
add_subdirectory(pybind11)

add_executable(main main.cpp)
target_link_libraries(main dl pybind11::embed pthread sndfile)
```


```C++
#include <pybind11/embed.h>
#include <sndfile.hh>
int main()
{
  sf_error(0);
  pybind11::scoped_interpreter guard{};
  pybind11::exec("import sklearn");
}
```

**Comment by jcelerier** [comment]
the assert passes on the python interpreter ; however I can't run it through pybind11 (it crashes at import sklearn)

**Comment by jcelerier** [comment]
I may have found a solution :

     sys.setdlopenflags(os.RTLD_NOW | os.RTLD_DEEPBIND)

**Comment by liversedge** [comment]
thank you @jcelerier I have also hit this same problem today -- that worked for me too on ubuntu 17.10

**Comment by pv** [comment]
These symbols however could be marked not exported (and generally so also for many other symbols), so that such workarounds are not needed.

**Comment by ffauvert** [comment]
From the "duplicated" and closed issue:  
> That the signatures differ does not matter when the shared libraries are loaded with the correct visibility settings. If you are embedding Python in your program, you can arrange this is so.

I would be happy to know how. Could you detail?  
Imagine the following scenario:  
* User launches a computation
* The computation raises an error which is output through Scipy through "int sf_error (SNDFILE *sndfile);" (scipy is loaded first so the sf_error in memory is the one of Scipy so, it's OK.
* After the output of the error, a sound is played to notify the user by a sound
* While playing the sound, an error is raised by libsndfile (file is corrupted, not accessible, sound card is not accessible, whatever)  
  
Don't you think it will raise a crash of the application? The system will not load the "sf_error" of the libsndfile because a symbol with the same name is already in memory.  
  
Please let me know how "If you are embedding Python in your program, you can arrange this is so".  
  
What you said in the other topic "These symbols however could be marked not exported (and generally so also for many other symbols), so that such workarounds are not needed." would be another real fix. Another one would be to rename "sf_error" to "scipy_sf_error" or something like that if symbol export is mandatory.

**Comment by tylerjereddy** [comment]
Should be fixed by #8463.
